### ISSUE TITLE
[[clang::lifetimebound]] false positive with template functions

### ISSUE THREAD

**Issue by fekir** [llvm/llvm-project#221223]
```c++
#include <span>

auto forward1(std::span<int> s [[clang::lifetimebound]]) { 
  return s; 
}

template <class T>
std::span<int> forward2(T&& s [[clang::lifetimebound]]) {
  return std::span<int>(s);
}

void bar(std::span<int> sp){
    {auto data = forward1(forward1(sp));} // success
    {auto data = forward1(forward2(sp));} // success
    {auto data =          forward2(sp) ;} // success
    {auto data = forward2(forward1(sp));} // fails with -Wdangling
    {auto data = forward2(forward2(sp));} // fails with -Wdangling
}
```

The same happens when replacing

```c++
template <class T>
std::span<int> forward2(T&& s [[clang::lifetimebound]]) {
  return std::span<int>(s);
}
```

with 

```c++
template <class T>
std::span<const int> forward2(const T& s [[clang::lifetimebound]]) {
  return std::span<const int>(s);
}
```
.

With

```c++
template <class T>
std::span<const int> forward2( std::span<T> s [[clang::lifetimebound]]) {
    return std::span<const int>(s);
}
```

no false positive is triggered.
